eis_toolkit/raster_processing/reclassify_raster.py

```diff
@@ -0,0 +1,484 @@
+            range_half = median_value - min_value
+            raster_half = raster_half - min_value + (range_half) / 1000.0
+
+        min_half = np.nanmin(raster_half)
+        max_half = np.nanmax(raster_half)
+
+        # number of classes
+        fac = (max_half / min_half) ** (1 / number_of_classes)
+
+        ibp = 1
+        brpt_half = [min_half]
+        brpt = [min_half]
+        width = [0]
+
+        while brpt[-1] < max_half:
+            ibp += 1
+            brpt.append(min_half * fac ** (ibp - 1))
+            brpt_half.append(brpt[-1])
+            width.append(brpt_half[-1] - brpt_half[0])
+        k = 0
+
+        for j in range(1, len(width) - 2):
+            values_out[
+                np.where(
+                    ((median_value + width[j]) < data_array)
+                    & (data_array <= (median_value + width[j + 1]))
+                    & (data_array != np.nan)
+                )
+            ] = (j + 1)
+            values_out[
+                np.where(
+                    ((median_value - width[j]) > data_array)
+                    & (data_array >= (median_value - width[j + 1]))
+                    & (data_array != np.nan)
+                )
+            ] = (-j - 1)
+            k = j
+
+        values_out[np.where(((median_value + width[k + 1]) < data_array) & (data_array != np.nan))] = k + 1
+        values_out[np.where(((median_value - width[k + 1]) > data_array) & (data_array != np.nan))] = -k - 1
+        values_out[np.where(median_value == data_array)] = 0
+
+        out_image.append(values_out)
+
+    return out_image, out_meta
+
+
+@beartype
+def raster_with_geometrical_intervals(  # type: ignore[no-any-unimported]
+    raster: rasterio.io.DatasetReader, number_of_classes: int, bands: Optional[Sequence[int]] = None
+) -> Tuple[Sequence[np.ndarray], dict]:
+    """Classify raster with geometrical intervals (Torppa, 2023).
+
+    If bands are not given, all bands are used for classification.
+
+    Args:
+        raster: Raster to be classified.
+        number_of_classes: The number of classes. The true number of classes is at most double the amount,
+        depending how symmetrical the input data is.
+        bands: Selected bands from multiband raster. Indexing begins from one. Defaults to None.
+
+    Returns:
+        Raster classified with geometrical intervals (Torppa, 2023) and metadata.
+    """
+    if bands is not None:
+        if not isinstance(bands, list):
+            raise InvalidParameterValueException("Expected bands parameter to be a list")
+        elif not all(isinstance(band, int) for band in bands):
+            raise InvalidParameterValueException("Expected bands to be a list of integers")
+        elif len(bands) > raster.count:
+            raise InvalidParameterValueException("The number of bands given exceeds the number of raster's bands")
+    if number_of_classes == 0:
+        raise InvalidParameterValueException("number_of_classes parameter is 0")
```

Comment:
I'd change this to something like this if negative values are not accepted:
```suggestion
    if number_of_classes <= 0:
        raise InvalidParameterValueException("The number of classes must be at least one")
```
Feel free to disagree.